Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp

```diff
@@ -0,0 +1,193 @@
+            // we need to remove it from the routeing table
+            // FIXME: This may be redundant due to being SSA like in our register usage
+            auto const& store = static_cast<Op::Store const&>(*it);
+            register_rerouteing_table.remove(store.dst().index());
+
+            break;
+        }
+        case DeleteById:
+        case DeleteByValue:
+            // These can trigger proxies, which call into user code
+            // So these are treated like calls
+        case GetByValue:
+        case GetById:
+        case PutByValue:
+        case PutById:
+            // Attribute accesses (`a.o` or `a[o]`) may result in calls to getters or setters
+            // May trigger Proxies
+            // So these are treated like calls
+        case Call:
+            // Calls, especially to local functions and eval, may poison visible and
+            // cached variables, hence we need to clear the lookup cache after emitting them
+            // FIXME: In strict mode and with better identifier metrics, we might be able
+            //        to safe some caching with a more fine-grained identifier table
+            // FIXME: We might be able to save some lookups to objects like `this`
+            //        which should not change their pointer
+            memcpy(new_block->next_slot(), &*it, (*it).length());
+            for (auto route : register_rerouteing_table)
+                reinterpret_cast<Instruction*>(new_block->next_slot())->replace_references(Register { route.key }, route.value);
+            new_block->grow((*it).length());
+
+            identifier_table.clear_with_capacity();
+
+            ++it;
+            continue;
+        case NewBigInt:
+            // FIXME: This is the only non trivially copyable Instruction,
+            //        so we need to do some extra work here
+            new (new_block->next_slot()) Op::NewBigInt(static_cast<Op::NewBigInt const&>(*it));
+            new_block->grow(sizeof(Op::NewBigInt));
+            ++it;
+            continue;
+        default:
+            break;
+        }
+
+        memcpy(new_block->next_slot(), &*it, (*it).length());
+        for (auto route : register_rerouteing_table) {
+            // Rerouteing from key to value
+            reinterpret_cast<Instruction*>(new_block->next_slot())->replace_references(Register { route.key }, route.value);
+        }
+        // because we are replacing the current block, we need to replace references
+        // to ourselves here
+        reinterpret_cast<Instruction*>(new_block->next_slot())->replace_references(block, *new_block);
+
+        new_block->grow((*it).length());
+
+        ++it;
+    }
+    return new_block;
+}
+
+void EliminateLoads::perform(PassPipelineExecutable& executable)
+{
+    started();
+
+    // FIXME: If we walk the CFG instead of the block list, we might be able to
+    //        save some work between blocks
```

Comment:
What's stopping us from doing that?

Reply:
Me not wanting to wrap my head around, when to restore or discard the cache, and the CFG not being 100% accurate in case of try-catch-finally handling.